rfcs/0182-user-management.md

```diff
@@ -0,0 +1,306 @@
+$> hometool generations remove 1-8
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Why not to keep this tool outside Nixpkgs monorepo?
+
+  It can be argued that overlays and other facilities provided by Nix language
+  allow to keep entire projects outside the Nixpkgs monorepo.
+
+  However, any of those projects suffer of an unpleasant phenomenon: the second
+  class citizen repo.
+
+  Basically, a repository outside the monorepo will not receive the same level
+  of care and attention of a project inside the monorepo. It happens because,
+  among other things,
+
+  - it is harder to pay attention in multiple repositories at the same time slot
+    than only one;
+  - it is harder to deal with pieces of code contained in distinct repositories;
+  - it is harder to synchronize multiple repositories;
+
+  Because of those hurdles, it becomes easier to ignore the repos outside the
+  main monorepo.
+
+  As a consequence, breaking changes in the monorepo are prone to be decided
+  without taking the satellite repos into consideration. This situation leads
+  the satellite repos to deal with the extra work of synchronizing with the
+  recent changes.
+
+  By living inside the main monorepo, the problems exposed above diminishes
+  drastically.
+
+  Further, having this home management toolset inside the monorepo brings many
+  advantages:
+
+  - No longer dealing with synchronization among projects
+  - Better strategies for code deduplication, sharing and refactoring
+  - Direct access to the most recent enhancements available
+
+    Indeed this is precisely what happens in NixOS already: when a program is
+    added, a NixOS module and test suite can be plugged in the same pull
+    request, with few to no bureaucracy.
+
+  - Reputation
+
+    A home management toolset kept in the monorepo has the social add-on of
+    being better regarded by the Nixpkgs community as a whole, since there will
+    be no barriers for contribution from this same community.
+
+- The monorepo will increase in size.
+
+  The increase in size is expected to be small.
+
+  Since we are building a set of modules to deal with the Nixpkgs packageset, a
+  good heuristic approximation of the worst case would be the current size of
+  `nixos` directory - after all, we want NixOS without superuser privileges.
+
+  A quick `du` returns something like this.
+
+  ```shell
+  $> du -sh nixos/ pkgs/
+  024572 nixos/
+  379536 pkgs/
+  ```
+
+  By the numbers above, obtained locally at 2024-09-13, `nixos` occupies less
+  than 7% of the sum of both `nixos + pkgs`. A quick and dirty calculation would
+  bring a similar 7% increase in size.
+
+  This is certainly a crude calculation, however we are ignoring many factors
+  that will bring these numbers down. Namely:
+
+  - Refactorings
+
+    Similar code can be factored and abstracted.
+
+  - Code sharing
+
+    Code that was initially projected for basic user management can be found
+    useful for superuser management too.
```

Comment:
There are many duplicated pieces of code between home-manager and nixos. 
Top of my mind is Chromium.

Reply:
Definitely! ...but that isn't what I'm asking here: it's *how* you would share code between a home-manager and NixOS module for Chromium. This is repeatedly mentioned in this RFC, but without any actual methods as to how to do it

The only one I can think of is [what we currently do](https://github.com/NixOS/nixpkgs/blob/6c8c52c5e0bcb76c914e9d641c7d64be860f24b4/nixos/modules/services/web-apps/pixelfed.nix#L97-L100) with NGINX's virtual host options a times, but using primitive language constructs like this for many shared options isn't a very good idea IMO. Again, this would probably need a good amount of work to specifically implement well -- unless you have some other ideas

Reply:
I do not expect this to be made methodically or automatically - my instinct suggest this is a problem on the fringes of undecidable.

In the specific case of Chromium, I can think of creating smaller modules for structuring things like
- command-line options
- extensions
- configuration file formats

In the end, the NixOS and Hometool modules would use the moduleset defined above and extend it according to  the particularities of their respective use cases.